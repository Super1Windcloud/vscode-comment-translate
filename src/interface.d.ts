import { Range } from "vscode";
import { StackElement, IToken, IGrammarExtensions } from "./syntax/TextMateService";

export interface ITokenState {
    startState: StackElement | null;
    tokens1: IToken[];
    endState: StackElement | null;
}

export interface IScopeLen {
    scopes: string[];
    len: number;
}

export interface ICommentToken {
    ignoreStart?: number;
    ignoreEnd?: number;
    text: string;
    scope?: IScopeLen[];
}

export interface ICommentOption {
    appRoot: string;
    grammarExtensions: IGrammarExtensions[];
    userLanguage: string;
}


export type checkScopeFunction = (scopes: string[]) => boolean;


export interface ICommentBlock {
	range: Range;
	comment: string;
	tokens?: ICommentToken[];
}



export interface ITranslatedText {
	translatedText: string;
	humanizeText?: string;
	targets: string[];
	texts:string[];
	combined:boolean[];
	translateLink: string;
}