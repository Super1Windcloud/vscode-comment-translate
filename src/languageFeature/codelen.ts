import type { CodeLensProvider, ProviderResult, TextDocument } from 'vscode'
import { CodeLens, EventEmitter, Range, languages } from 'vscode'

class PlaceholderCodeLensProvider implements CodeLensProvider {
  public linesOfDoc = new Map<string, Set<number>>()

  changeCodeLensesEmitter = new EventEmitter<void>()

  onDidChangeCodeLenses = this.changeCodeLensesEmitter.event

  provideCodeLenses(document: TextDocument): ProviderResult<CodeLens[]> {
    const codeLens: CodeLens[] = []
    const lines = this.linesOfDoc.get(document.uri.toString())
    if (!lines)
      return codeLens

    lines.forEach((line) => {
      codeLens.push(new CodeLens(new Range(
        line,
        0,
        line,
        0,
      ), {
        title: ' ',
        command: '',
        arguments: [document],
      }))
    })
    return codeLens
  }
}

let instance: PlaceholderCodeLensProvider

function refresh() {
  instance.changeCodeLensesEmitter.fire()
}


function append(document:TextDocument, rangeList:number[]) {
  const uri = document.uri.toString()
  let lines = instance.linesOfDoc.get(uri)
  if (!lines) {
    lines = new Set();
    instance.linesOfDoc.set(uri, lines);
  }

  rangeList.forEach((line) => {
    lines!.add(line);
  })

  refresh();
}

function cleanAll() {
  instance.linesOfDoc.clear()
  refresh()
}

export function usePlaceholderCodeLensProvider() {
  if (!instance) {
    instance = new PlaceholderCodeLensProvider()
    languages.registerCodeLensProvider({ scheme: 'file' }, instance)
  }

  return {
    append,
    cleanAll,
  }
}
