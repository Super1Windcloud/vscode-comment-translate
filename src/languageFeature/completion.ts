
import { CompletionItem, CompletionItemKind, CompletionItemProvider, ExtensionContext, languages, Position, ProviderResult, TextDocument } from "vscode";
import { getVariableCompletions, setCaseGuide } from "../command/replaceSelections";

class VarTranslateCompletionItemProvider implements CompletionItemProvider {
    async provideCompletionItems(document: TextDocument, position: Position) {
        // 不需要触发字符
        // if(context.triggerCharacter) return null;
        let coms = getVariableCompletions(document, new Position(position.line, position.character));
        if(!coms) {
            return null;
        }


        let {range, list,filterText,caseGuide,codeType} = coms;
        let r = range;
        
        let res: CompletionItem[] = list.map((c, index) => {
            let item = new CompletionItem(c.value);
            item.kind = convertKind(codeType);
            item.range = r;
            item.filterText = filterText;
            item.insertText = c.value;
            item.sortText = `${index}`;
            item.detail = `${c.title}`;
            item.documentation = "This is a variable completion item";

            if(caseGuide[codeType] === c.title) {
                item.preselect = true;
            }

            return item;
        });

        return res;
    }

    resolveCompletionItem(item: CompletionItem): ProviderResult<CompletionItem> {
        item.kind&&setCaseGuide(convertCodeType(item.kind), item.detail!);
        return item;
    }
}

export function registerCompletion(context: ExtensionContext, canLanguages: string[] = []) {
    const completionItemProviderDisposable = languages.registerCompletionItemProvider(canLanguages, new VarTranslateCompletionItemProvider());
    context.subscriptions.push(completionItemProviderDisposable);
}



function convertKind(codeType:string) {
    switch(codeType) {
        case 'variable':
            return CompletionItemKind.Variable;
        case 'function':
            return CompletionItemKind.Function;
        case 'class':
            return CompletionItemKind.Class;
        case 'readonly':
            return CompletionItemKind.Constant;
        default:
            return CompletionItemKind.Text;
    }
}

function convertCodeType(kind:CompletionItemKind) {
    switch(kind) {
        case CompletionItemKind.Variable:
            return 'variable';
        case CompletionItemKind.Function:
            return 'function';
        case CompletionItemKind.Class:
            return 'class';
        case CompletionItemKind.Constant:
            return 'readonly';
        default:
            return 'variable';
    }
}