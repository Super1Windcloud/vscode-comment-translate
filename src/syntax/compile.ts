import humanizeString = require("humanize-string");
import { getConfig } from "../configuration";
import { autoMutualTranslate, translateManager } from "../translate/manager";
import { hasEndMark, isLowerCase, isUpperCase } from "../util/string";
import { ICommentBlock, ICommentToken, ITranslatedText } from "../interface";



function ignoreStringTag(tokens: ICommentToken[], regular: string) {
    if (regular) {
        return tokens.map(item => {
            let { ignoreStart = 0, ignoreEnd = 0, text } = item;
            const validText = text.slice(ignoreStart, text.length - ignoreEnd);
            let match = validText.match('^' + regular);
            if (match && match.length) {
                ignoreStart += match[0].length;
            }
            item.ignoreStart = ignoreStart;

            let endMatch = validText.match('\\s+$');
            if (endMatch && endMatch.length) {
                ignoreEnd += endMatch[0].length;
            }
            item.ignoreEnd = ignoreEnd;

            return item;
        });
    }
    return tokens;
}

function humanize(originText: string) {
    const needHumanize = originText.trim().indexOf(' ') < 0;
    if (needHumanize) {
        // Converted to natural language
        return humanizeString(originText);
    }
    return '';
}

function combineLine(texts: string[]) {
    let combined: boolean[] = []; // 标记被合并行。 便于翻译后重新组合
    let combinedTexts = texts.reduce<string[]>((prev, curr, index) => {
        let lastIndex = combined.lastIndexOf(false);
        combined[index] = false;
        if (prev.length > 0) {
            let last = prev[lastIndex];
            if (isUpperCase(last) && hasEndMark(last) && isLowerCase(curr)) {
                // If it can be merged, merge to the previous row
                prev[lastIndex] = last + ' ' + curr;
                //The current line is blank, but the blank placeholder remains
                curr = '';
                combined[index] = true;
            }
        }
        prev.push(curr);
        return prev;
    }, []);

    return { combined, combinedTexts };
}


function getIgnoreRegular(languageId: string) {
    const ignore = getConfig<{ languageId: string, regular: string }[]>('ignore');
    if (!ignore) return '';
    let { regular = '' } = ignore.find(item => {
        return item.languageId.split(',').some((text => text.trim() === languageId));
    }) || {};
    return regular;
}

export async function compileBlock(block: ICommentBlock, languageId: string, targetLanguage?: string): Promise<ITranslatedText> {

    let translatedText: string;
    let targets: string[] = [];
    let texts: string[] = [];
    let combined: boolean[] = []; // Marked as merged rows for easy regrouping after translation
    let humanizeText: string = '';
    const { comment: originText } = block;
    let { tokens } = block;

    if (!tokens) {
        // No tokens means select translation or single word translation, only need to produce simple results
        humanizeText = humanize(originText);
        translatedText = await autoMutualTranslate(humanizeText || originText, { to: targetLanguage });
    } else {
        // Tokens represent comments, strings, and need to be structured

        // Regular ignore partial structure content
        let regular = getIgnoreRegular(languageId) || '[\\s|/]+';
        tokens = ignoreStringTag(tokens, regular);

        // Get the string to be translated.
        texts = tokens.map(({ text, ignoreStart = 0, ignoreEnd = 0 }) => {
            return text.slice(ignoreStart, text.length - ignoreEnd).trim();
        });

        // When multiline merge is enabled, multiple lines in a valid string are merged into the same line.
        if (getConfig<boolean>('multiLineMerge')) {
            let res = combineLine(texts);
            combined = res.combined;
            texts = res.combinedTexts;
        }

        // Filter blank lines to solve the problem of partial translation source, multi-line blank compression.
        let validTexts = texts.filter(text => {
            return text.length > 0;
        });
        let validText = validTexts.join('\n');
        let validTextLen = validText.length;

        // When there is no string to translate, the empty string is displayed directly, skipping the translation process.
        if (validTextLen === 0) {
            translatedText = originText;
        } else {
            // 只有1行，并且符合大小切换
            if (tokens.length === 1) {
                humanizeText = humanize(validText);
            }
            translatedText = await autoMutualTranslate(humanizeText || validText, { to: targetLanguage });

            // Reassemble the translation results to restore the filtered matches when translated, such as/* //, etc.
            targets = translatedText.split('\n');
            if (translatedText && validTexts.length === targets.length) {
                let translated = [];
                for (let i = 0, j = 0; i < tokens.length; i++) {
                    const { text, ignoreStart = 0, ignoreEnd = 0 } = tokens[i];
                    const translateText = texts[i];
                    let targetText = '';
                    if (translateText.length > 0) {
                        targetText = targets[j];
                        j += 1;
                    }
                    // Merged rows skipped
                    if (targetText === '' && combined[i]) {
                        continue;
                    }
                    const startText = text.slice(0, ignoreStart);
                    const endText = text.slice(text.length - ignoreEnd);
                    translated.push(startText + targetText + endText);
                }
                translatedText = translated.join('\n');
            }
        }
    }

    return {
        translatedText,
        humanizeText,
        targets,
        texts,
        combined,
        translateLink: translateManager.link(humanizeText || originText)
    };
}
