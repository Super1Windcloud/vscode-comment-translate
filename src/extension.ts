/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';

import { ExtensionContext, extensions, env, window } from 'vscode';
import { registerCommands } from './command/command';
import { mouseToSelect } from './command/select';
import { getConfig, onConfigChange, showHoverStatusBar, showTargetLanguageStatusBarItem } from './configuration';
import { registerDefinition } from './languageFeature/definition';
import { registerHover } from './languageFeature/hover';
import { AliTranslate } from './plugin/translateAli';
import { BingTranslate } from './translate/BingTranslate';
import { GoogleTranslate } from './translate/GoogleTranslate';
import { ITranslateConfig, ITranslateRegistry, TranslateExtensionProvider } from './translate/translateExtension';
import { ITranslateOptions, TranslateManager } from 'comment-translate-manager';
import { Comment } from './syntax/Comment';
import { TextMateService } from './syntax/TextMateService';
import { commentDecorationManager } from './languageFeature/decoration';
import { extractGrammarExtensions, readResources } from './util/ext';
import { detectLanguage } from './lang';
import { registerCompletion } from './languageFeature/completion';
import { initTranslate } from './translate/manager';

export let outputChannel = window.createOutputChannel('Comment Translate');
export let comment: Comment;
let canLanguages: string[] = ['plaintext'];

export let userLanguage: string;

export let ctx: ExtensionContext;

export async function activate(context: ExtensionContext) {
    ctx = context;

    // let languageId = 2;
    let { grammarExtensions, languageId } = extractGrammarExtensions([...extensions.all], 2);
    // 如果为远程环境，使用插件内置语法
    if (env.remoteName) {
        const inner = await readResources(context.extensionPath);
        let { grammarExtensions: innergrammarExtensions } = extractGrammarExtensions(inner, languageId);
        grammarExtensions.push(...innergrammarExtensions);
    }

    canLanguages = grammarExtensions.reduce<string[]>(((prev, item) => {
        let lang: string[] = item.value.map((grammar) => grammar.language).filter(v => v);
        return prev.concat(lang);
    }), canLanguages);


    let BlackLanguage: string[] = ['log', 'Log', 'code-runner-output'];
    canLanguages = canLanguages.filter((v) => BlackLanguage.indexOf(v) < 0);
    userLanguage = env.language;

    let langMaps: Map<string, string> = new Map([
        ['zh-cn', 'zh-CN'],
        ['zh-tw', 'zh-TW'],
    ]);
    // 修复语言代码不一致
    if (langMaps.has(userLanguage)) {
        userLanguage = langMaps.get(userLanguage) || '';
    }

    const textMate = new TextMateService(grammarExtensions);
    comment = new Comment(textMate);
    let translateManager = initTranslate(context, userLanguage);

    translateManager.onTranslate(e => {
        outputChannel.append(e);
    });


    registerCommands(context);
    registerHover(context, canLanguages);
    registerDefinition(context, canLanguages);
    registerCompletion(context, canLanguages);

    context.subscriptions.push(...commentDecorationManager.showBrowseCommentTranslate(canLanguages));
    // 注册状态图标
    let hoverBar = await showHoverStatusBar();
    let targetBar = await showTargetLanguageStatusBarItem(userLanguage);
    context.subscriptions.push(targetBar, hoverBar, outputChannel, comment);
    mouseToSelect(context);

    // Exposing Translation Plugins
    return {
        extendTranslate: function (registry: ITranslateRegistry) {
            registry('ali.cloud', AliTranslate);
        }
    }
}





