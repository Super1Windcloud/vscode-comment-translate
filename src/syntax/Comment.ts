import { Disposable, ExtensionContext, Position, Range, TextDocument, workspace } from "vscode";
import { CommentParse } from "./CommentParse";
import { TextMateService } from "./TextMateService";
import { ICommentBlock } from "../interface";
import { getGrammerExtensions } from "../util/ext";
import { ctx } from "../extension";

export class Comment implements Disposable {

    private _disposable: Disposable;
    private _commentParseCache: Map<string, CommentParse> = new Map();
    constructor(private _textMateService: TextMateService) {
        // Close document or content changes, remove cache
        this._disposable = Disposable.from(
            workspace.onDidChangeTextDocument(e => this._removeCommentParse(e.document))
        );
    }

    dispose() {
        this._commentParseCache.clear();
        this._disposable?.dispose();
    }

    _removeCommentParse({ languageId, uri }: TextDocument) {
        const key = `${languageId}-${uri}`;
        this._commentParseCache.delete(key);
    }

    //缓存已匹配部分，加快hover运行时间
    async _getCommentParse(textDocument: TextDocument) {
        const { uri, languageId } = textDocument;
        const key = `${languageId}-${uri}`;
        if (this._commentParseCache.has(key)) {
            return this._commentParseCache.get(key);
        }
        const grammar = await this._textMateService.createGrammar(languageId);
        if (grammar == null)
            return null;
        const parse = new CommentParse(textDocument, grammar);
        parse.maxLineLength = workspace.getConfiguration('editor').get('maxTokenizationLineLength', 20000) as number;
        this._commentParseCache.set(key, parse);
        return parse;
    }

    async getComment(textDocument: TextDocument, position: Position): Promise<ICommentBlock | null> {
        const parse = await this._getCommentParse(textDocument);
        if (!parse) return null;
        return parse.computeText(position);
    }

    async getAllComment(textDocument: TextDocument, type = 'comment', range?: Range): Promise<ICommentBlock[] | null> {
        const parse = await this._getCommentParse(textDocument);
        if (!parse) return null;
        return parse.computeAllText(type, range);
    }

    async getWordAtPosition(textDocument: TextDocument, position: Position) {
        const parse = await this._getCommentParse(textDocument);
        if (!parse) return null;
        return parse.getWordAtPosition(position);
    }

}

let comment: Comment;
export async function createComment() {
    if (comment) return comment;
    let grammarExtensions = await getGrammerExtensions();
    const textMate = new TextMateService(grammarExtensions);
    comment = new Comment(textMate);

    ctx.subscriptions.push(comment);
    return comment;
}
