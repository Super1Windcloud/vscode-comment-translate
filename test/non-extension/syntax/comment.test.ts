
jest.mock("../../../src/configuration", () => {
  return {
    getConfig: jest.fn(),
  };
});

import {translateManager} from '../../../src/extension';
import { Comment } from "../../../src/syntax/Comment";
import { Uri } from "vscode";
import { createTextDocument } from "jest-mock-vscode";
import { codeData } from "../../fixtures/codes";
import { compileBlock } from "../../../src/languageFeature/compile";
import { getConfig } from '../../../src/configuration';
import {  mockTextMateService } from '../mocks';


jest.mock("../../../src/extension", ()=>{
  return {
    translateManager: {
      translate: jest.fn(),
      link: jest.fn()
    }
}
});

let ignoreConfig = [
  {
    "languageId": "javascript,typescript,javascriptreact,typescriptreact",
    "regular": "[\\*\\s]+"
  },
  {
    "languageId": "dart",
    "regular": "[\\s|/]+"
  },
  {
    "languageId": "rust",
    "regular": "[\\s|/]+"
  }
];


describe("Comment", () => {
  let comment: Comment;
  beforeAll(async () => {
    let textMateService = await mockTextMateService();
    comment = new Comment(textMateService);
  });

  test.each(codeData)("Documents should be compiled comment correctly:【$name】", async (item) => {
    let uri = Uri.file(item.name);
    let curr_doc = createTextDocument(uri, item.code, item.languageId);
    let blocks = await comment.getAllComment(curr_doc, "comment");

    (getConfig as jest.Mock).mockImplementation((key: string) => {
      switch (key) {
        case "multiLineMerge":
          return item.multiLineMerge;
        case "ignore":
          return ignoreConfig;
        default:
          return;
      }
    });

    expect(blocks).toMatchSnapshot();
    // 注释翻译区域，需要同实际翻译内容区域一致。确定拆分块正常
    expect(blocks?.length).toEqual(item.translated.length);

    if (blocks && blocks.length > 0) {
      let blocksTasks = blocks.map(async (block, index) => {
        (translateManager.translate as jest.Mock).mockResolvedValueOnce(item.translated[index]);
        let res = await compileBlock(block, item.languageId);
        expect(res).toMatchSnapshot();
      });

      await Promise.all(blocksTasks);
    }

  });
});
