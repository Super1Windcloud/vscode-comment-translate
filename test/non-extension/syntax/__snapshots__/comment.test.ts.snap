// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Documents should be compiled comment correctly:【function comment】 1`] = `
[
  {
    "comment": "/**
	 * Represents an event describing the change in a {@link TextEditor.options text editor's options}.
	 */",
    "range": [
      {
        "character": 0,
        "line": 0,
      },
      {
        "character": 4,
        "line": 2,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 3,
        "text": "/**",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 3,
        "text": "	 * Represents an event describing the change in a {@link TextEditor.options text editor's options}.",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 4,
        "text": "	 */",
      },
    ],
  },
  {
    "comment": "/**
		 * The new value for the {@link TextEditor.options text editor's options}.
		 */",
    "range": [
      {
        "character": 2,
        "line": 6,
      },
      {
        "character": 5,
        "line": 8,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 3,
        "text": "/**",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 4,
        "text": "		 * The new value for the {@link TextEditor.options text editor's options}.",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 5,
        "text": "		 */",
      },
    ],
  },
]
`;

exports[`Comment Documents should be compiled comment correctly:【head comment】 1`] = `
[
  {
    "comment": "// Type definitions for Visual Studio Code 1.68
// Project: https://github.com/microsoft/vscode
// Definitions by: Visual Studio Code Team, Microsoft <https://github.com/microsoft>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped",
    "range": [
      {
        "character": 0,
        "line": 0,
      },
      {
        "character": 66,
        "line": 3,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "// Type definitions for Visual Studio Code 1.68",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "// Project: https://github.com/microsoft/vscode",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "// Definitions by: Visual Studio Code Team, Microsoft <https://github.com/microsoft>",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped",
      },
    ],
  },
  {
    "comment": "/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License.
 *  See https://github.com/microsoft/vscode/blob/main/LICENSE.txt for license information.
 *--------------------------------------------------------------------------------------------*/",
    "range": [
      {
        "character": 0,
        "line": 5,
      },
      {
        "character": 96,
        "line": 9,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 95,
        "text": "/*---------------------------------------------------------------------------------------------",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 1,
        "text": " *  Copyright (c) Microsoft Corporation. All rights reserved.",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 1,
        "text": " *  Licensed under the MIT License.",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 1,
        "text": " *  See https://github.com/microsoft/vscode/blob/main/LICENSE.txt for license information.",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 96,
        "text": " *--------------------------------------------------------------------------------------------*/",
      },
    ],
  },
  {
    "comment": "/**
 * Type Definition for Visual Studio Code 1.68 Extension API
 * See https://code.visualstudio.com/api for more information
 */",
    "range": [
      {
        "character": 0,
        "line": 11,
      },
      {
        "character": 3,
        "line": 14,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 3,
        "text": "/**",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": " * Type Definition for Visual Studio Code 1.68 Extension API",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": " * See https://code.visualstudio.com/api for more information",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 3,
        "text": " */",
      },
    ],
  },
]
`;

exports[`Comment Documents should be compiled comment correctly:【spaceInHead comment】 1`] = `
[
  {
    "comment": "// This is the theme of your application
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.",
    "range": [
      {
        "character": 0,
        "line": 0,
      },
      {
        "character": 28,
        "line": 8,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "// This is the theme of your application",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        //",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // Try running your application with "flutter run". You'll see the",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // application has a blue toolbar. Then, without quitting the app, try",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // changing the primarySwatch below to Colors.green and then invoke",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // "hot reload" (press "r" in the console where you ran "flutter run",",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // or simply save your changes to "hot reload" in a Flutter IDE).",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // Notice that the counter didn't reset back to zero; the application",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        // is not restarted.",
      },
    ],
  },
  {
    "comment": "/// This is the theme of your application
        ///
        /// Try running your application with "flutter run". You'll see the
        /// application has a blue toolbar. Then, without quitting the app, try
        /// changing the primarySwatch below to Colors.green and then invoke
        /// "hot reload" (press "r" in the console where you ran "flutter run",
        /// or simply save your changes to "hot reload" in a Flutter IDE).
        /// Notice that the counter didn't reset back to zero; the application
        /// is not restarted.",
    "range": [
      {
        "character": 8,
        "line": 12,
      },
      {
        "character": 29,
        "line": 20,
      },
    ],
    "tokens": [
      {
        "ignoreEnd": 0,
        "ignoreStart": 2,
        "text": "/// This is the theme of your application",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        ///",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// Try running your application with "flutter run". You'll see the",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// application has a blue toolbar. Then, without quitting the app, try",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// changing the primarySwatch below to Colors.green and then invoke",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// "hot reload" (press "r" in the console where you ran "flutter run",",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// or simply save your changes to "hot reload" in a Flutter IDE).",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// Notice that the counter didn't reset back to zero; the application",
      },
      {
        "ignoreEnd": 0,
        "ignoreStart": 10,
        "text": "        /// is not restarted.",
      },
    ],
  },
]
`;
